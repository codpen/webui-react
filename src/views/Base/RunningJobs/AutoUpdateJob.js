import React from 'react';
import axiosInstance from "../../../utils/API/API";
import PropTypes from "prop-types";
import {Alert} from "reactstrap";

const propTypes = {
    jobId: PropTypes.number.isRequired,
    jobCompleteHandler: PropTypes.func.isRequired,
    cardDismissedHandler: PropTypes.func.isRequired
};


class AutoUpdateJob extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            job: {},
            isOpen: true
        };
        this.loadJobs = this.loadJobs.bind(this);
    }

    async loadJobs() {
        try {
            let res = await axiosInstance.post("/job/status", {jobid: this.props.jobId});
            console.log("jobStatus", res.data);

            this.setState({job: res.data});

            if (this.state.job.finished) {
                this.props.jobCompleteHandler();
                clearInterval(this.jobRefreshInterval)
            }

        } catch (e) {
            // console.log(`Error loading jobs: core/stats ${e}`);
            console.log("Job is complete");

            // this.props.jobCompleteHandler();

        }
    }

    componentDidMount() {
        this.loadJobs();
        this.jobRefreshInterval = setInterval(this.loadJobs, 1000);
    }

    componentWillUnmount() {
        clearInterval(this.jobRefreshInterval);
    }

    dismissAlert = () => {
        this.setState({isOpen: !this.state.isOpen});
        this.props.cardDismissedHandler();
    };


    render() {
        return (
            <Alert color="info" isOpen={this.state.isOpen} toggle={this.dismissAlert} sm={12}
                   lg={12}>
                {/*<Progress value={percentage} className={"mb-2"}>{percentage} %</Progress> /!*percentage*!/*/}
            </Alert>
        );


    }
}

AutoUpdateJob.propTypes = propTypes;

export default AutoUpdateJob;
