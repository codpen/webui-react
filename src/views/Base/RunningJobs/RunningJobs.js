import React from 'react';
import {Card, CardBody, CardHeader, Col, Progress, Row} from "reactstrap";
import {bytesToMB, formatBytes, secondsToStr} from "../../../utils/Tools";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {Line} from "react-chartjs-2";
import {CustomTooltips} from "@coreui/coreui-plugin-chartjs-custom-tooltips";

const options = {
    tooltips: {
        enabled: false,
        custom: CustomTooltips
    },
    maintainAspectRatio: false
}
function JobCard({job}) {
    const {name, eta, percentage, speed, speedAvg, size, bytes} = job;
    return (<Card>
        <CardHeader>Running Jobs</CardHeader>
        <CardBody>
            <p>{name}</p> {/*Name of the file*/}
            <Progress value={percentage} className={"mb-2"}>{percentage} %</Progress> {/*percentage*/}
            <p><strong>Speed: </strong>{formatBytes(speed)}PS</p> {/*speed*/}
            <p><strong>Average Speed: </strong>{formatBytes(speedAvg)}PS</p> {/*speedAvg*/}
            <p><strong>Total transferred: </strong>{formatBytes(bytes)}</p> {/*bytes: convert to mb*/}
            <p><strong>Size: </strong>{formatBytes(size)}</p>
            <p><strong>ETA: </strong>{secondsToStr(eta)} seconds</p>
        </CardBody>

    </Card>);
}

function JobCardRow({job}) {
    const {name, percentage, speed, size} = job;
    return (
        <React.Fragment>
            <Row>
                {(size && speed) ? (<Col lg={12}>{name}({formatBytes(size)}) - {formatBytes(speed)}PS </Col>) : (
                    <Col lg={12}>Calculating</Col>)}

            </Row>
            <Row>
                <Col lg={12}><Progress value={percentage} className={"mb-2"}>{percentage} %</Progress></Col>
            </Row>

        </React.Fragment>
    );

}

function GlobalStatus({stats}) {
    const {speed, bytes, checks, elapsedTime, deletes, errors, transfers} = stats;
    return (
        <Card>
            <CardHeader><strong>Global Stats</strong></CardHeader>
            <CardBody>
                <p><strong>Bytes Transferred: </strong>{formatBytes(bytes)}</p>
                <p><strong>Average Speed: </strong>{formatBytes(speed)}PS</p>
                <p><strong>Checks: </strong>{checks}</p>
                <p><strong>Deletes: </strong>{deletes}</p>
                <p><strong>Running since: </strong>{secondsToStr(elapsedTime)}</p>
                <p><strong>Errors: </strong>{errors}</p>
                <p><strong>Transfers: </strong>{transfers}</p>

            </CardBody>
            {/*<CardFooter></CardFooter>*/}

        </Card>);

}

function TransferringJobs({transferring}) {
    if (transferring !== undefined) {
        return transferring.map((item, idx) => {
            return (<JobCard key={idx} job={item}/>);
        });
    }
    return null;
}

function TransferringJobsRow({transferring}) {
    if (transferring !== undefined) {
        return transferring.map((item, idx) => {
            return (<JobCardRow key={idx} job={item}/>);
        });
    }
    return null;
}


class RunningJobs extends React.Component {


    render() {
        const {jobs, isConnected, lineChartData} = this.props;
        const {transferring} = jobs;
        const {mode} = this.props;
        if (mode === "full-status") {
            if (isConnected) {
                return (
                    <Row>
                        <Col sm={12} lg={4}>
                            <GlobalStatus stats={jobs}/>
                        </Col>
                        <Col sm={12} lg={4}>
                            <TransferringJobs transferring={transferring}/>
                        </Col>
                        <Col sm={12} lg={4}>
                            <Card>
                                <CardHeader>
                                    Speed
                                    <div className="card-header-actions">
                                        <a href="http://www.chartjs.org" className="card-header-action">
                                            <small className="text-muted">docs</small>
                                        </a>
                                    </div>
                                </CardHeader>
                                <CardBody>
                                    <div className="chart-wrapper">
                                        <Line data={lineChartData} options={options}/>
                                    </div>
                                </CardBody>
                            </Card>
                        </Col>
                    </Row>
                );
            } else {
                return (<div>Not connected to rclone.</div>)
            }

        } else if (mode === "card") {
            if (isConnected) {
                return (

                    <TransferringJobsRow transferring={transferring}/>
                );
            } else {
                return (<div>Not connected to rclone.</div>);
            }

        } else if (mode === "modal") {
            if (transferring && transferring.length > 0)
                return (
                    <Card className={"progress-modal"}>
                        <CardHeader>Progress</CardHeader>
                        <CardBody>
                            <TransferringJobsRow transferring={transferring}/>

                        </CardBody>
                    </Card>
                );
            return null;
        }
    }
}

RunningJobs.propTypes = {
    mode: PropTypes.string.isRequired,
    isConnected: PropTypes.bool.isRequired,
    jobs: PropTypes.object.isRequired,
    error: PropTypes.object
};

const mapStateToProps = (state, ownProps) => {

    const speedData = state.status.speed;
    let lineChartData = {};
    if (speedData) {
        let labels = [];
        let data = [];

        const dataLength = speedData.length;
        //
        const limitedData = speedData.slice(dataLength - 100, dataLength - 1);
        // console.log(limitedData.length);
        limitedData.forEach((item, idx) => {
            labels.push(Math.ceil(item.elapsedTime));
            data.push(bytesToMB(item.speed).toFixed(2));
        });

        lineChartData = {
            labels: labels,
            datasets: [
                {
                    label: 'Average Speed (mbps)',
                    fill: false,
                    lineTension: 0.1,
                    backgroundColor: 'rgba(75,192,192,0.4)',
                    borderColor: 'rgba(75,192,192,1)',
                    borderCapStyle: 'butt',
                    borderDash: [],
                    borderDashOffset: 0.0,
                    borderJoinStyle: 'miter',
                    pointBorderColor: 'rgba(75,192,192,1)',
                    pointBackgroundColor: '#fff',
                    pointBorderWidth: 1,
                    pointHoverRadius: 5,
                    pointHoverBackgroundColor: 'rgba(75,192,192,1)',
                    pointHoverBorderColor: 'rgba(220,220,220,1)',
                    pointHoverBorderWidth: 2,
                    pointRadius: 1,
                    pointHitRadius: 10,
                    data: data,
                },
            ],
        };
    }


    return {
        jobs: state.status.jobs,
        isConnected: state.status.isConnected,
        error: state.status.error,
        lineChartData
    }
};

export default connect(mapStateToProps, {})(RunningJobs);
