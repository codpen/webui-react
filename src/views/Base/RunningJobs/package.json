{
  "name": "RunningJobs",
  "version": "0.0.0",
  "private": true,
  "main": "RunningJobs.js"
}
