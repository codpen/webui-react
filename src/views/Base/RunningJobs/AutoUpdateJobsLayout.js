import React from "react";
import AutoUpdateJob from "./AutoUpdateJob";
import axiosInstance from "../../../utils/API/API";

class AutoUpdateJobsLayout extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            jobs: []
        }
        this.refreshInterval = null;
        this.getJobs = this.getJobs.bind(this);
    }

    // addJob = (jobId) => {
    //     const jobs = this.state.jobs;
    //     if (jobs.indexOf(jobId) === -1) {
    //         jobs.push(jobId);
    //     }
    //
    // };

    jobCompleteHandler = (jobId) => {
        console.log("Job complete: " + jobId);
    };

    cardDismissedHandler = (jobId) => {
        console.log("Card dismissed: " + jobId);
    };

    async getJobs() {
        let res = await axiosInstance.post("/job/list", {});
        console.log(res);
        const currentJobs = this.state.jobs;
        const jobList = res.data.jobids;

        jobList.map((item, idx) => {
            if (currentJobs.indexOf(item) === -1) {
                currentJobs.push(item);
            }
        });
        this.setState({jobs: currentJobs});
    }

    componentDidMount() {
        if (this.refreshInterval === null) {
            this.refreshInterval = setInterval(() => this.getJobs(), 2000);
        }
    }

    render() {
        return (
            <React.Fragment>
                {
                    this.state.jobs.map((item, idx) => {
                        return (<AutoUpdateJob key={idx} jobId={item} jobCompleteHandler={this.jobCompleteHandler}
                                               cardDismissedHandler={this.cardDismissedHandler}/>)
                    })
                }
            </React.Fragment>
        );
    }
}

export default AutoUpdateJobsLayout;