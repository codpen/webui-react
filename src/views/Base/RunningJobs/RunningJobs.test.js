import React from "react";
import {shallow} from "enzyme";
import {testStore} from "../../../../Utils";
import RunningJobs from "./RunningJobs";

const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);
    const component = shallow(<RunningJobs {...props} store={store}/>);
    return component.childAt(0).dive();
};


describe('Running Jobs', function () {


    describe('renders', function () {
        let wrapper;
        beforeEach(() => {
            const initialState = {
                status: {
                    jobs: {
                        "transferring": {}
                    },
                    isConnected: true,
                    error: {}
                }
            };

            const props = {
                mode: "full-status"
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            expect(wrapper).toHaveLength(1)
        });


    });




});