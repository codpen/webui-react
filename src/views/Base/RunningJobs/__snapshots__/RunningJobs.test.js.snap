// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Running Jobs renders should match snapshot 1`] = `
<Row
  tag="div"
>
  <Col
    lg={4}
    sm={12}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <GlobalStatus
      stats={
        Object {
          "transferring": Object {},
        }
      }
    />
  </Col>
  <Col
    lg={4}
    sm={12}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <TransferringJobs
      transferring={Object {}}
    />
  </Col>
  <Col
    lg={4}
    sm={12}
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Card
      tag="div"
    >
      <CardHeader
        tag="div"
      >
        Speed
        <div
          className="card-header-actions"
        >
          <a
            className="card-header-action"
            href="http://www.chartjs.org"
          >
            <small
              className="text-muted"
            >
              docs
            </small>
          </a>
        </div>
      </CardHeader>
      <CardBody
        tag="div"
      >
        <div
          className="chart-wrapper"
        >
          <Line
            data={Object {}}
            options={
              Object {
                "maintainAspectRatio": false,
                "tooltips": Object {
                  "custom": [Function],
                  "enabled": false,
                },
              }
            }
          />
        </div>
      </CardBody>
    </Card>
  </Col>
</Row>
`;
